### ISSUE TITLE
bot is too slow to confirm when picking up loot

### ISSUE THREAD

**Issue by leewheel** [liyunfan1223/mod-playerbots#626]
**Describe the bug**
In the dungeon, the bot is too slow to confirm when picking up loot, and sometimes it doesn't confirm whether it is demand or greed

**Commit hash**
The hash of the current commit.

**To Reproduce**
Just LFG and test

**Comment by hermensbas** [comment]
Which hardware cpu and mem usage and how many bots do you have configured?

**Comment by leewheel** [comment]
13th Gen Intel(R) Core(TM) i5-13500HX   2.50 GHz  32.0 GB Memory，use about 6GB, 1000bots

**Comment by hermensbas** [comment]
hmm interesting will try and reproduce

**Comment by leewheel** [comment]
Thanks！

**Comment by xPh03n1x** [comment]
I just recently started toying around with AC and PlayerBots. I've been experiencing this as well to the point where I got annoyed today and started looking for a solution.

I am encountering this everywhere, it's not just in dungeons.

I have the server running on a 8 core Intel(R) Core(TM) i7-6700, with 64 GBs of RAM. The databases are serviced by MySQL 8.0, fine-tuned for large amount of traffic via consecutive connections (I use very similar configuration in a production environment for the platforms of a large corp).

I run the server with random number of bots, between 200 and 4000. While trying to get more information about this today, I lowered the maximum bots to 50 and I still waited for 19 minutes for a Greed roll from my party bots on a Malachite. I stopped the server after that, didn't even get the item. Malachite is not even an item they should be calculating item-level-upgrade logic on.

I am using the very latest version of PlayerBots and nearly no additional modules.

If there are any additional details I can try to extract to help get this resolved - please, let me know.

For the time being I am just setting `AiPlayerbot.LootNeedRollLevel` to 0, hoping that will resolve this.

P.S.
I realize this issue has been Closed but no fix ever seems to have been pushed. If I should open a new issue, let me know and I will do so.

**Comment by YunataSavior** [comment]
Two things related to need/greed and playerbots:
1. I'd like for bots to always default to greed. Do we have a config for this?
2. It does take a really long time for bots to need/greed/pass from my observation, so we should address that here.

**Comment by Dohtt** [comment]
From my experience they need/greed or pass after combat ended or after they had their drink in a dungeon.

18 minutes is impossible and wouldn't have to do with the bot behavior (as once it times out, it should auto pass no matter if everyone selected an option or not.

**Comment by xPh03n1x** [comment]
> 1. I'd like for bots to always default to greed. Do we have a config for this?

From the `playerbots.conf`
```
# Bots' Roll level bots will use for items they Need (0 = pass, 1 = greed, 2 = need)
# Default: 1 (greed)
AiPlayerbot.LootNeedRollLevel = 0
```
**HOWEVER** I just played for a couple of hours with my wife, doing dungeons and questing with bots. This setting absolutely does not apply to all bots and all items.
We had 1 bot in a dungeon always doing Greed, the other 2 were passing (as expected with this setting). A couple times a locked box dropped (those that need a rogue to unlock them) - all bots rolled Greed (even though the setting says they should Pass on it).
Whilst questing we had a tanking bot - that boot rolled greed on everything but Mail items, on them it rolled Need (again, all the while this setting is set to `0` and the bot should be passing on everything)



> From my experience they need/greed or pass after combat ended or after they had their drink in a dungeon.
> 
> 18 minutes is impossible and wouldn't have to do with the bot behavior (as once it times out, it should auto pass no matter > if everyone selected an option or not.

The 19 minutes example I gave was from a super straight forward case, no complications or edge case scenarios: login, attack a bear with 4 player bots (not in a dungeon); kill the bear; loot; roll on Malachite; wait for 19 minutes; shut down the server.

While I understand your logic for the timeout, I have to disagree on the assumption this isn't related to playerbots. Last week every night after work I played with colleagues on this server with no bots in our group. We did a ton of dungeons and never encountered a problem with the Need/Greed system. On Sunday I started playing with bots so I can see how they work and I started seeing this quite often.


I should also mention that I tried tinkering with the settings
```
# Max AI iterations per tick
AiPlayerbot.IterationsPerTick = 50

# Delay between two bot actions
AiPlayerbot.ReactDelay = 10
```
but that doesn't seem to have any effect on this.

I am also using these values for the threads:
```
PlayerbotsDatabase.WorkerThreads     = 4
PlayerbotsDatabase.SynchThreads     = 4
```
The server is connected to the database via socket, not TCP, so that eliminates any possible packet interruptions as well.

Even though my hardware and MySQL settings are perfectly capable of handling all the requests and resource consumption, tomorrow I will do some tests with significantly less bots (I'm thinking no more than 100?).

**Comment by Celandriel** [comment]
I wonder if the packet is being dropped for some reason

**Comment by xPh03n1x** [comment]
In a dungeon with bots with my laptop tailing the Playerbots.log file - loot drop, no one votes, this keeps repeating:

```
Letina --- AI Tick ---
Letina T:loot available
Letina PUSH:loot - 6.000000 (trigger)
Letina PUSH:follow - 1.000000 (default)
Letina A:loot - PREREQ
Letina A:loot - FAILED
Letina A:follow - USELESS
Letina no actions executed
Kainis --- AI Tick ---
Kainis T:loot available
Kainis PUSH:loot - 6.000000 (trigger)
Kainis PUSH:follow - 1.000000 (default)
Kainis PUSH:pull action - 105.000000 (default)
Kainis A:pull action - USELESS
Kainis A:loot - PREREQ
Kainis A:loot - FAILED
Kainis A:follow - USELESS
Kainis no actions executed
Kainis --- AI Tick ---
Kainis T:loot available
Kainis PUSH:loot - 6.000000 (trigger)
Kainis PUSH:follow - 1.000000 (default)
Kainis PUSH:pull action - 105.000000 (default)
Kainis A:pull action - USELESS
Kainis A:loot - PREREQ
Kainis A:loot - FAILED
Kainis A:follow - USELESS
Kainis no actions executed
Shalii --- AI Tick ---
Shalii T:far from current loot
Shalii PUSH:move to loot - 7.000000 (trigger)
Shalii PUSH:follow - 1.000000 (default)
Shalii A:move to loot - PREREQ
Shalii A:move to loot - FAILED
Shalii A:follow - USELESS
Shalii no actions executed
Letina --- AI Tick ---
Letina T:loot available
Letina PUSH:loot - 6.000000 (trigger)
Letina PUSH:follow - 1.000000 (default)
Letina A:loot - PREREQ
Letina A:loot - FAILED
Letina A:follow - USELESS
Letina no actions executed
Kainis --- AI Tick ---
Kainis T:loot available
Kainis PUSH:loot - 6.000000 (trigger)
Kainis PUSH:follow - 1.000000 (default)
Kainis PUSH:pull action - 105.000000 (default)
Kainis A:pull action - USELESS
Kainis A:loot - PREREQ
Kainis A:loot - FAILED
Kainis A:follow - USELESS
Kainis no actions executed
Kainis --- AI Tick ---
Kainis T:loot available
Kainis PUSH:loot - 6.000000 (trigger)
Kainis PUSH:follow - 1.000000 (default)
Kainis PUSH:pull action - 105.000000 (default)
Kainis A:pull action - USELESS
Kainis A:loot - PREREQ
Kainis A:loot - FAILED
Kainis A:follow - USELESS
Kainis no actions executed
Letina --- AI Tick ---
Letina T:loot available
Letina PUSH:loot - 6.000000 (trigger)
Letina PUSH:follow - 1.000000 (default)
Letina A:loot - PREREQ
Letina A:loot - FAILED
Letina A:follow - USELESS
Letina no actions executed
```
Those were the bots in my party. Not sure if that helps, maybe it shows you something that could suggest where we can look ?

**Comment by xPh03n1x** [comment]
I found the reason why this happens to me (can't be sure about anyone else) - literally the only module I use, [mod-aoe-loot](https://github.com/azerothcore/mod-aoe-loot).

When **mod-aoe-loot** is compiled in - there's problem with the Need/Greed system and bots don't vote on gear.

After recompiling with the module removed I did about 7 dungeons and a couple of hours questing - never had a problem with the Need/Greed votes. Sometimes bots votes a bit slower but they still vote.

Recompiled with **mod-aoe-loot**, entered a dungeon, encountered the problem on the second item vote.

I recompiled with and without a couple of more times to validate and it happened every time: no problem without **mod-aoe-loot**, problem when compiled with.

Looking at the module's code, I think this relates to the item mailing system they wrote.

At this point I think this can be safely closed as not-related to Playerbots.

**Comment by ZarkXav** [comment]
> In a dungeon with bots with my laptop tailing the Playerbots.log file - loot drop, no one votes, this keeps repeating:
> 
> ```
> Letina --- AI Tick ---
> Letina T:loot available
> Letina PUSH:loot - 6.000000 (trigger)
> Letina PUSH:follow - 1.000000 (default)
> Letina A:loot - PREREQ
> Letina A:loot - FAILED
> Letina A:follow - USELESS
> Letina no actions executed
> Kainis --- AI Tick ---
> Kainis T:loot available
> Kainis PUSH:loot - 6.000000 (trigger)
> Kainis PUSH:follow - 1.000000 (default)
> Kainis PUSH:pull action - 105.000000 (default)
> Kainis A:pull action - USELESS
> Kainis A:loot - PREREQ
> Kainis A:loot - FAILED
> Kainis A:follow - USELESS
> Kainis no actions executed
> Kainis --- AI Tick ---
> Kainis T:loot available
> Kainis PUSH:loot - 6.000000 (trigger)
> Kainis PUSH:follow - 1.000000 (default)
> Kainis PUSH:pull action - 105.000000 (default)
> Kainis A:pull action - USELESS
> Kainis A:loot - PREREQ
> Kainis A:loot - FAILED
> Kainis A:follow - USELESS
> Kainis no actions executed
> Shalii --- AI Tick ---
> Shalii T:far from current loot
> Shalii PUSH:move to loot - 7.000000 (trigger)
> Shalii PUSH:follow - 1.000000 (default)
> Shalii A:move to loot - PREREQ
> Shalii A:move to loot - FAILED
> Shalii A:follow - USELESS
> Shalii no actions executed
> Letina --- AI Tick ---
> Letina T:loot available
> Letina PUSH:loot - 6.000000 (trigger)
> Letina PUSH:follow - 1.000000 (default)
> Letina A:loot - PREREQ
> Letina A:loot - FAILED
> Letina A:follow - USELESS
> Letina no actions executed
> Kainis --- AI Tick ---
> Kainis T:loot available
> Kainis PUSH:loot - 6.000000 (trigger)
> Kainis PUSH:follow - 1.000000 (default)
> Kainis PUSH:pull action - 105.000000 (default)
> Kainis A:pull action - USELESS
> Kainis A:loot - PREREQ
> Kainis A:loot - FAILED
> Kainis A:follow - USELESS
> Kainis no actions executed
> Kainis --- AI Tick ---
> Kainis T:loot available
> Kainis PUSH:loot - 6.000000 (trigger)
> Kainis PUSH:follow - 1.000000 (default)
> Kainis PUSH:pull action - 105.000000 (default)
> Kainis A:pull action - USELESS
> Kainis A:loot - PREREQ
> Kainis A:loot - FAILED
> Kainis A:follow - USELESS
> Kainis no actions executed
> Letina --- AI Tick ---
> Letina T:loot available
> Letina PUSH:loot - 6.000000 (trigger)
> Letina PUSH:follow - 1.000000 (default)
> Letina A:loot - PREREQ
> Letina A:loot - FAILED
> Letina A:follow - USELESS
> Letina no actions executed
> ```
> 
> Those were the bots in my party. Not sure if that helps, maybe it shows you something that could suggest where we can look ?

I had a similar issue. I tried using AI to resolve the issue, but I really don't know what's going on tbh. I tried the entire repack 32 module Azeroth one, compiled my own multiple times with and without different modules and code to attempt to fix it. I'd really like to fix this issue, and I'm just waiting on an update from playerbots because I assume something must be wrong at this point? Is there any possible fixes?

**Comment by ZarkXav** [comment]
> > In a dungeon with bots with my laptop tailing the Playerbots.log file - loot drop, no one votes, this keeps repeating:
> > ```
> > Letina --- AI Tick ---
> > Letina T:loot available
> > Letina PUSH:loot - 6.000000 (trigger)
> > Letina PUSH:follow - 1.000000 (default)
> > Letina A:loot - PREREQ
> > Letina A:loot - FAILED
> > Letina A:follow - USELESS
> > Letina no actions executed
> > Kainis --- AI Tick ---
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Those were the bots in my party. Not sure if that helps, maybe it shows you something that could suggest where we can look 


I had a similar issue. I tried using AI to resolve the issue, but I really don't know what's going on tbh. I tried the entire repack 32 module Azeroth one, compiled my own multiple times with and without different modules and code to attempt to fix it. I'd really like to fix this issue, and I'm just waiting on an update from playerbots because I assume something must be wrong at this point? Is there any possible fixes?

**Comment by jplrobinson-cyber** [comment]
Same issue here, and removing mod-aoe-loot did not fix the problem.
